Add flag to preserve cached media on cleanup

northerner opened the pull request:
Adding the flag --keep-interacted to the "media remove" tootctl command to preserve media cached from other servers if the related status has interactions from local accounts.

For example, you may wish to delete all cached media from other servers older than 3 months to save disk space on the local server. However, if the status was bookmarked by a user they likely want to keep a copy locally, preserving the media in case the remote server is closed or has decided themselves to clear out all old user media.

This flag was proposed in issue #30449.

The change does not impact the local content retention setting, which manages automated cleanup. A new setting would need to be added there in a future improvement.

This means to make use of this change a server admin would have to set a long content retention setting beyond which all remote media would be cleared, and then use a shorter range with the tootctl command where interacted media would be preserved with this new flag.

The flag is only to preserve media attached to statuses, and makes no difference to the logic removing cached avatars and headers.

Commit:
Add flag to preserve cached media on cleanup

Adding the flag --keep-interacted to the "media remove" tootctl command
to preserve media cached from other servers if the related status has
interactions from local accounts.

For example, you may wish to delete all cached media from other servers
older than 3 months to save disk space on the local server. However, if
the status was bookmarked by a user they likely want to keep a copy
locally, preserving the media in case the remote server is closed or has
decided themselves to clear out all old user media.

This flag was proposed in issue #30449.

The change does not impact the local content retention setting, which
manages automated cleanup. A new setting would need to be added there in
a future improvement.

This means to make use of this change a server admin would have to set a
long content retention setting beyond which all remote media would be
cleared, and then use a shorter range with the tootctl command where
interacted media would be preserved with this new flag.

The flag is only to preserve media attached to statuses, and makes no
difference to the logic removing cached avatars and headers.

ClearlyClaire reviewed (changes requested):
This is unfortunately far too slow to be useful in practice.

I'm trying to rewrite the query in a better way, but this is very tricky to do considering how complex it is and the fact it involves joining on the same table multiple times.

ClearlyClaire commented:
I think I would have a scope directly on `MediaAttachment` instead of `Status` to avoid the double-join, and it would read something like:
```ruby
  scope :without_local_interaction, -> {
    where.not(Favourite.joins(:account).merge(Account.local).where(Favourite.arel_table[:status_id].eq(MediaAttachment.arel_table[:status_id])).select(1).arel.exists)
    .where.not(Bookmark.where(Bookmark.arel_table[:status_id].eq(MediaAttachment.arel_table[:status_id])).select(1).arel.exists)
    .where.not(Status.local.where(Status.arel_table[:in_reply_to_id].eq(MediaAttachment.arel_table[:status_id])).select(1).arel.exists)
    .where.not(Status.local.where(Status.arel_table[:reblog_of_id].eq(MediaAttachment.arel_table[:status_id])).select(1).arel.exists)
  }
```

jeroenhabets commented on lib/mastodon/cli/media.rb:
```diff
@@ -17,6 +17,7 @@ class UnrecognizedOrphanType < StandardError; end
     option :concurrency, type: :numeric, default: 5, aliases: [:c]
     option :verbose, type: :boolean, default: false, aliases: [:v]
     option :dry_run, type: :boolean, default: false
+    option :keep_interacted, type: :boolean, default: false
```
@northerner, @ClearlyClaire assuming you agree `keep_interacted` is the sensible default and the current design does not enable that. Wouldn't it be better to have a `remove-interacted` instead?

```
 If --remove-interacted is specified, any media attached to a status that
      was favourited, bookmarked, replied to, or reblogged by a local account
      will also be removed.
```

Commit:
Refactor interaction scope to attachment model

Using a more efficient scope, suggested by @ClearlyClaire on the
MediaAttachment model to filter attachments without local server
interaction.

Commit:
Account for quotes in interacted media scope

northerner commented on app/models/media_attachment.rb:
```diff
@@ -214,6 +214,12 @@ class MediaAttachment < ApplicationRecord
+    where.not(Favourite.joins(:account).merge(Account.local).where(Favourite.arel_table[:status_id].eq(MediaAttachment.arel_table[:status_id])).select(1).arel.exists)
+         .where.not(Bookmark.where(Bookmark.arel_table[:status_id].eq(MediaAttachment.arel_table[:status_id])).select(1).arel.exists)
+         .where.not(Status.local.where(Status.arel_table[:in_reply_to_id].eq(MediaAttachment.arel_table[:status_id])).select(1).arel.exists)
+         .where.not(Status.local.where(Status.arel_table[:reblog_of_id].eq(MediaAttachment.arel_table[:status_id])).select(1).arel.exists)
```
Thanks @ClearlyClaire, those extra clauses for this scope work great. Now the scope will preserve media that was in a status quoted by a local account, as well as media on any remote status that is quoting a local account.

Updated the tests with extra cases.
